### PR TITLE
k8s_secret: new module to create secrets

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

create secrets on kubernetes analogous to ``kubectl create secret``

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

k8s_secret
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by abikouo** [comment]
recheck

File: changelogs/fragments/352-k8s_secret-module.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+major_changes:
+  - k8s_secret - New module for creating secrets, analogous to `kubectl create secret` (https://github.com/ansible-collections/community.kubernetes/pull/352).
```
Comment (Akasurde) [inline changelogs/fragments/352-k8s_secret-module.yml:3]: I think, there is no need for this file for the new module.

File: molecule/default/library/k8s_molecule_x509_cert.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021, Aubin Bikouo <@abikouo>
```
Comment (Akasurde) [inline molecule/default/library/k8s_molecule_x509_cert.py:3]:
```suggestion
# Copyright: (c) 2022, Aubin Bikouo <@abikouo>
```

File: plugins/action/k8s_info.py [hunk 3]
```diff
@@ -263,16 +274,38 @@ def get_file_realpath(self, local_path):
...
+        if self._task.action not in (
+            "k8s_secret",
+            "kubernetes.core.k8s_secret",
+            "community.okd.k8s_secret",
+            "redhat.openshift.k8s_secret",
+            "community.kubernetes.k8s_secret",
+        ):
             raise AnsibleActionFail(
-                "%s does not exist in local filesystem" % local_path
+                "'tls_certificate' and 'from_path' are only supported parameters for the 'k8s_secret' module."
```
Comment (Akasurde) [inline plugins/action/k8s_info.py:291]:
```suggestion
                "'tls_certificate' and 'from_path' are only supported parameters for the 'kubernetes.core.k8s_secret' module."
```

File: plugins/modules/k8s_secret.py [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+module: k8s_secret
+
+short_description: Create, update or delete a secret
+
+version_added: "2.3.0"
+
+author: Aubin Bikouo (@abikouo)
+
+description:
+  - Use Kubernetes Python SDK to create generic, tls or docker-registry secrets on Kubernetes.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:23]:
```suggestion
  - Use Kubernetes Python SDK to create C(generic), C(tls), or C(docker-registry) secrets on Kubernetes.
```

File: plugins/modules/k8s_secret.py [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+  - Analogous to `kubectl create secret`.
+  - A generic secret can be created using a file, directory, or specified literal value.
+  - When creating a generic secret based on a file, the key will default to the basename of the file, and the value will default to
+    the file content. If the basename is an invalid key or you wish to chose your own, you may specify an alternate key.
+  - When creating a generic secret based on a directory, each file whose basename is a valid key in the directory will be packaged
+    into the secret. Any directory entries except regular files are ignored (e.g. subdirectories, symlinks, devices, pipes, etc).
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:29]:
```suggestion
    into the secret. Any directory entries except regular files are ignored (for example, subdirectories, symlinks, devices, pipes, and so on).
```

File: plugins/modules/k8s_secret.py [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+  - A TLS secret is created from a given public/private key pair.
+    The public/private key pair must exist before hand. The public key certificate must be .PEM encoded and match the given private key.
+  - A docker-registry secret is used to authenticate against Docker registries.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:32]:
```suggestion
  - A C(docker-registry) secret is used to authenticate against Docker registries.
```

File: plugins/modules/k8s_secret.py [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+    - Append a hash of the secret to its name.
+    type: bool
+    default: false
+  type:
+    description:
+    - Specifies the type of Secret to create.
+    - When set to I(generic) one of C(literals) or C(from_path) should be set.
+    - When set to I(docker-registry) one of C(docker-registry-credentials), C(literals) or C(from_path) should be set.
+    - Ignored when C(state) is set to I(absent).
+    - PyOpenSSL python library is required to create tls secret.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:61]:
```suggestion
    - PyOpenSSL python library is required to create C(tls) secret.
```

File: plugins/modules/k8s_secret.py [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+    required: true
+    type: str
+  namespace:
+    description:
+      - Use to specify a Secret object namespace.
+    required: true
+    type: str
+  docker_registry_credentials:
+    description:
+      - Provide details for docker registry authentication.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:80]:
```suggestion
      - Provide details for Docker registry authentication.
```

File: plugins/modules/k8s_secret.py [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+    description:
+      - Specifies file or directory defining key and value to insert in secret.
+      - Key files can be specified using their file path, in which case a default name will be given to
+        them, or as dictionnary with a name and file path, in which case the given name will be used.
+      - Specifying a directory will iterate each named file in the directory that is a valid secret key.
+      - Ignored when C(type) is not set to I(generic).
+    type: raw
+  tls_certificate:
+    description:
+      - Provide tls certificate along with corresponding private key to create a new tls secret.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:114]:
```suggestion
      - Provide TLS certificate along with the corresponding private key to create a new C(tls) secret.
```

File: plugins/modules/k8s_secret.py [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+    type: dict
+    suboptions:
+        certificate:
+            description:
+            - Path to PEM encoded public key certificate.
+            type: path
+            required: true
+        private_key:
+            description:
+            - Path to private key associated with given certificate.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:126]:
```suggestion
            - Path to the private key associated with the given certificate.
```

File: plugins/modules/k8s_secret.py [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+            type: path
+            required: true
+  literals:
+    description:
+      - dictionnary of key and literal value to insert in secret data.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:131]:
```suggestion
      - Dictionary of key and literal value to insert in secret data.
```

File: plugins/modules/k8s_secret.py [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+    description:
+      - Specifies the type of generic secret to create.
+      - When set to I(kubernetes.io/service-account-token) you need to ensure that the C(annotations) contains
+        key 'kubernetes.io/service-account.name' set to an existing service account name. The module may report that
+        the secret has been created when the service account does not exist, but it is not.
+      - Ignores when C(type) is not set to I(generic).
+    type: str
+    default: Opaque
+    choices:
+    - Opaque
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:144]:
```suggestion
    - opaque
```
Can we make this all lowercase just like other options?

File: plugins/modules/k8s_secret.py [hunk 13]
```diff
@@ -0,0 +1,724 @@
...
+    - kubernetes.io/service-account-token
+    - kubernetes.io/dockerconfigjson
+    - kubernetes.io/basic-auth
+    - kubernetes.io/ssh-auth
+    - bootstrap.kubernetes.io/token
+  labels:
+    description:
+    - dictionnary of string keys and values used as labels for the secret to be created.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:152]:
```suggestion
    - A dictionary of string keys and values is used as labels for the secret to be created.
```

File: plugins/modules/k8s_secret.py [hunk 14]
```diff
@@ -0,0 +1,724 @@
...
+    - see U(http://kubernetes.io/docs/user-guide/labels).
+    - Ignored if C(state) is set to I(absent).
+    type: dict
+  annotations:
+    description:
+    - dictionnary of string keys and values used as annotations for the secret to be created.
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:158]:
```suggestion
    - A dictionary of string keys and values is used as annotations for the secret to be created.
```

File: plugins/modules/k8s_secret.py [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+- name: Create a new TLS secret
+  kubernetes.core.k8s_secret:
+    name: my-tls-secret
+    namespace: my-secret-namespace
+    type: tls
+    tls_certificate:
+        certificate: path/to/tls.key
+        private_key: path/to/tls.cert
+
+- name: Create a  docker configuration secret
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:211]:
```suggestion
- name: Create a Docker configuration secret
```

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: plugins/modules/k8s_secret.py (hunk 12 above)
Comment (abikouo) [reply plugins/modules/k8s_secret.py:144]: ok I will do

File: changelogs/fragments/352-k8s_secret-module.yml (hunk 1 above)
Comment (abikouo) [reply changelogs/fragments/352-k8s_secret-module.yml:3]: ok thanks will be removed

File: plugins/modules/k8s_secret.py [hunk 16]
```diff
@@ -0,0 +1,724 @@
...
+            description:
+            - Email for Docker registry.
+            type: str
+  from_path:
+    description:
+      - Specifies file or directory defining key and value to insert in secret.
+      - Key files can be specified using their file path, in which case a default name will be given to
+        them, or as dictionnary with a name and file path, in which case the given name will be used.
+      - Specifying a directory will iterate each named file in the directory that is a valid secret key.
+      - Ignored when C(type) is not set to I(generic).
```
Comment (Akasurde) [inline plugins/modules/k8s_secret.py:110]:
```suggestion
    description:
      - Specifies a file or directory defining key and value to insert in secret.
      - Key files can be specified using their file path, in which case a default name will be given to
        them, or a dictionary with a name and file path, in which case the given name will be used.
      - Specifying a directory will iterate each named file in the directory that is a valid secret key.
      - Ignored when C(type) is not set to I(generic).
```

**Review by Akasurde — APPROVED** [review]

**Comment by abikouo** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Comment by gravesm** [comment]
Thanks for your work on this. I'm not sure if this is something we want to merge at this time, though. We have generally tried to avoid creating modules for specific kubernetes resources.

**Comment by abikouo** [comment]
> Thanks for your work on this. I'm not sure if this is something we want to merge at this time, though. We have generally tried to avoid creating modules for specific Kubernetes resources.

I am getting the point but this is currently done for the service with the module ``k8s_service`` and today users are forced to use ``kubectl`` executable if they want to create a secret, so I think this is definitely something which will be helpful for this collection.

**Comment by gravesm** [comment]
`k8s_service` is an outlier and something I would actually prefer to try and eventually remove. You can still create a secret with the `k8s` module, and this would be easy to do if your secrets were stored in vault or some other secrets management solution, which seems like the more likely scenario. Part of the problem with this approach is that we're creating two different ways of managing the same k8s resource, making it more likely to have divergent behavior.

**Pull request closed without merging**